### PR TITLE
Add typings for version 3.0.0

### PR DESCRIPTION
**Pull request by EndBug**
As a follow up to issue #211, I made typings for the new version of the package. I've tried to make them as much readable and maintainable as I could, also including the methods not listed in the docs and the interfaces from the Google API.
I may have missed something, so a second check would be great :)

### CODE REVIEW

**Comment by theoephraim** [comment]
Couple questions:
- is including `@types/node` necessary? Doesn't seem to be used
- what does the tsconfig file do in this case where you are just adding types?

**Comment by EndBug** [comment]
> is including `@types/node` necessary? Doesn't seem to be used 

Yeah, sorry, I just forgot to remove it. Fixed ✔️ 

> what does the tsconfig file do in this case where you are just adding types?

The `tsconfig` doesn't have any effect on the types when they're imported from the package. The only reason why it's there is that when editing the types you need to tell the IDE that you're working on a TS project, overwise it will not add corrections. It can be removed, but then every time someone has to edit the type they would also need to add another one to get type-checking.

@theoephraim

**Comment by theoephraim** [comment]
gotcha. Thanks :)

One more thing - maybe it would be helpful to break this up a bit. Particularly keeping the enums/interfaces that are defined by google and not likely to ever change in another file? Is it common to break up class type definitions into their own files? I'm just trying to think about how to keep this the most easily maintainable for the future.

I suppose the other option would be to convert the project itself to use TS and then much of this would be defined inline?

**Comment by EndBug** [comment]
Breaking up the file would be a good idea, but I've searched a bit and there doesn't seem to be any easy way to do that: everything I've found includes either not declaring this as a module (which doesn't seem like a good idea to me) or declaring everything as a module. I think it would be way easier to just keep it there.

The other option would be switching to TypeScript. That would be awesome because you could use those types too when making edits to the library, and it could prevent a lot of stupid stuff from happening. I personally love using TypeScript and I switched to it in almost all of my projects.
There's something to keep in mind though: you'll need to recompile the library every time you want to publish a new version, since you have to publish a JS version to npm. That is pretty easy since you can set up a GitHub Actions workflow that keeps the build updated and also publishes the package every time you bump the version. Types will be generated automatically along with the compiled version of the library.

If you decide to switch to TS I'll be happy to help :)

**Comment by theoephraim** [comment]
OK thanks - I'll do a little research and see what I can find :)

File: typings/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+    setHeaderRow(headerValues: string[]): Promise<void>
+    addRow(values: object): Promise<GoogleSpreadsheetRow>
+    getRows(options?: {
+      offset?: number
+      limit?: number
+    }): Promise<GoogleSpreadsheetRow>
+
+    // Working With Cells
+    loadCells(filters?: any): Promise<any>
+    getCells(rowIndex: number, columnIndex: number): GoogleSpreadsheetCell
```
Comment (vogre) [inline typings/index.d.ts]:
```suggestion
    getCell(rowIndex: number, columnIndex: number): GoogleSpreadsheetCell
```

Looks like a typo compared to [doc](https://theoephraim.github.io/node-google-spreadsheet/#/classes/google-spreadsheet-worksheet?id=fn-getcell)

File: typings/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+    constructor(parentSpreadsheet: GoogleSpreadsheet, { properties, data })
+
+    // Working With Rows
+    loadHeaderRow(): Promise<void>
+    setHeaderRow(headerValues: string[]): Promise<void>
+    addRow(values: object): Promise<GoogleSpreadsheetRow>
+    getRows(options?: {
+      offset?: number
+      limit?: number
+    }): Promise<GoogleSpreadsheetRow>
```
Comment (vogre) [inline typings/index.d.ts]:
```suggestion
    }): Promise<GoogleSpreadsheetRow[]>
```

The result is expected to be an array in [doc](https://theoephraim.github.io/node-google-spreadsheet/#/classes/google-spreadsheet-worksheet?id=fn-getrows).

**Comment by EndBug** [comment]
Thanks, I've added the changes :)

**Comment by theoephraim** [comment]
closing now since we're going to merge the other one. But may use this as the basis for migrating to TS. Thanks again!!

**Pull request closed without merging**